[ao] QConfigMapping generation for Quantization Config ModelReport

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83358
* #83454

Summary: This adds the capability to generate a qconfigmapping for the
modules that have suggestions as given in the ModelReport API.

Currently, this adds the capability to generate a qconfigmapping that
gives suggesstions about modules about:
- Whether dynamic vs static quantization should be used
- Whether per_channel or per_tensor quantization should be used

For the other modules that aren't applicable to any of these, the
default qconfig is used.

This qconfigmapping that is compiled is returned to the user to use for
preparing and converting their model.

This feature has to happen after report generation as this is when the
statistics are gathered by the detectors.

Test Plan: python test/test_quantization.py TestFxModelReportClass.test_qconfig_mapping_generation

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] QConfigMapping generation for Quantization Config ModelReport

Summary: This adds the capability to generate a qconfigmapping for the
modules that have suggestions as given in the ModelReport API.

Currently, this adds the capability to generate a qconfigmapping that
gives suggesstions about modules about:
- Whether dynamic vs static quantization should be used
- Whether per_channel or per_tensor quantization should be used

For the other modules that aren't applicable to any of these, the
default qconfig is used.

This qconfigmapping that is compiled is returned to the user to use for
preparing and converting their model.

This feature has to happen after report generation as this is when the
statistics are gathered by the detectors.

Test Plan: python test/test_quantization.py TestFxModelReportClass.test_qconfig_mapping_generation

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1146,15 +1146,58 @@ def test_qconfig_mapping_generation(self):
+
+            # try to generate mapping without generating report, should throw error
+            # only really need to test one of them here
+            with self.assertRaises(Exception):
+                qconfig_mapping = mod_report.generate_qconfig_mapping()
```
is it possible to generate the qconfig_mapping from the report object? so that we don't need to do these checks

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report.py:
```diff
@@ -6,12 +6,24 @@
     DETECTOR_IS_POST_OBS_KEY,
-    DETECTOR_TARGET_NODE_KEY
+    DETECTOR_TARGET_NODE_KEY,
+    DynamicStaticDetector,
+    PerChannelDetector
```
I feel model_report should not depend on individual detectors...

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report.py:
```diff
@@ -436,7 +448,76 @@ def generate_qconfig_mapping(self) -> QConfigMapping:
+
+        # now we create a high level mapping with a default qconfig
+        return_mapping = QConfigMapping()
+        return_mapping.set_global(default_qconfig)
```
do we want to set a global default qconfig?

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report.py:
```diff
@@ -436,7 +448,76 @@ def generate_qconfig_mapping(self) -> QConfigMapping:
+        return_mapping.set_global(default_qconfig)
+
+        # get the set of modules that are support per channel
+        current_backend = torch.backends.quantized.engine
```
also I feel it might make more sense to ask user for the backend, the quantized engine when we call this function may not be the same as the quantized engine during actual inference

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report.py:
```diff
@@ -6,12 +6,24 @@
+    default_weight_observer,
+)
 
 class ModelReport:
```
I think ideally, each detector will be able to generate a qconfig_mapping and the job for model_reporter class is to merge them in some way, this will make the code more modular, is this possible?

vspenubarthi committed:
Update on "[ao] QConfigMapping generation for Quantization Config ModelReport"


Summary: This adds the capability to generate a qconfigmapping for the
modules that have suggestions as given in the ModelReport API.

Currently, this adds the capability to generate a qconfigmapping that
gives suggesstions about modules about:
- Whether dynamic vs static quantization should be used
- Whether per_channel or per_tensor quantization should be used

For the other modules that aren't applicable to any of these, the
default qconfig is used.

This qconfigmapping that is compiled is returned to the user to use for
preparing and converting their model.

This feature has to happen after report generation as this is when the
statistics are gathered by the detectors.

Test Plan: python test/test_quantization.py TestFxModelReportClass.test_qconfig_mapping_generation

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] QConfigMapping generation for Quantization Config ModelReport"


Summary: This adds the capability to generate a qconfigmapping for the
modules that have suggestions as given in the ModelReport API.

Currently, this adds the capability to generate a qconfigmapping that
gives suggesstions about modules about:
- Whether dynamic vs static quantization should be used
- Whether per_channel or per_tensor quantization should be used

For the other modules that aren't applicable to any of these, the
default qconfig is used.

This qconfigmapping that is compiled is returned to the user to use for
preparing and converting their model.

This feature has to happen after report generation as this is when the
statistics are gathered by the detectors.

Test Plan: python test/test_quantization.py TestFxModelReportClass.test_qconfig_mapping_generation

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
